### PR TITLE
[sled-agent] Destroy VNICs in a background task

### PR DESCRIPTION
**Pull request by smklein**
- Implements a `Destructor` struct in `illumos-utils`, which helps asynchronously destroy objects. This should be a general tool in cases where...
  - You have an object that you want to destroy when it is dropped...
  - ... but destroying the object may be (1) asynchronous, (2) fallible, or (3) both.
- This `Destructor` is then added to the `VnicAllocator`, where it is referenced by created VNICs.
  - As a result: when VNICs are dropped, they are handled by a background task, until they are successfully destroyed.

Fixes https://github.com/oxidecomputer/omicron/issues/2656

### CODE REVIEW

File: illumos-utils/src/link.rs [hunk 1]
```diff
@@ -178,29 +207,42 @@ impl Link {
...
-            eprintln!("Failed to delete Link: {}", e);
+        if let Some(destructor) = self.destructor.take() {
+            destructor
+                .enqueue_destroy(LinkDestruction { name: self.name.clone() });
         }
     }
 }
 
+// Represents the request to destroy a VNIC
+struct LinkDestruction {
```
Comment (bnaecker) [inline illumos-utils/src/link.rs:218]: Not sure if we'll delete other kinds of links, but if so, consider renaming to `VnicDestruction`

File: illumos-utils/src/link.rs [hunk 2]
```diff
@@ -124,28 +155,24 @@ pub struct InvalidLinkKind(String);
...
-            Some(kind) => Ok(Self {
-                name: name.as_ref().to_owned(),
-                deleted: false,
-                kind,
-            }),
-            None => Err(InvalidLinkKind(name.as_ref().to_owned())),
-        }
+impl std::fmt::Debug for Link {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
```
Comment (bnaecker) [inline illumos-utils/src/link.rs:167]: Nit: I tend to use [`Formatter::debug_struct`](https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_struct) when I want to just limit the debug output to a subset of fields.

File: illumos-utils/src/destructor.rs [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+            // Unwrap safety:
+            // - The DestructorWorker can only be stopped by "try_close"
+            // - "try_close" can only succeed if it's called on the last strong reference to the
+            // Destructor.
+            // - Since "enqueue_destroy" was called on a "Destructor", one more reference must
+            // exist.
+            .unwrap();
+    }
+
+    /// Closes the destructor if this is the only reference to it.
```
Comment (bnaecker) [inline illumos-utils/src/destructor.rs:159]: Maybe make a note that `Self` is returned in the `Result::Err` if there is more than one reference outstanding.

File: illumos-utils/src/destructor.rs [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+        let mut exit = false;
+        loop {
+            tokio::select! {
+                _ = self.futs.next() => {
+                    if exit && self.futs.is_empty() {
+                        return;
+                    }
+                }
+                msg = self.rx.recv() => {
+                    let msg = match msg {
```
Comment (bnaecker) [inline illumos-utils/src/destructor.rs:64]: Maybe `let msg = msg.unwrap_or(Message::Exit)`?

File: illumos-utils/src/destructor.rs [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+        // Note that the strong reference inside "obj"
+        // also counts as a strong reference.
+        let destructor = Destructor::<ObjectDestruction>::new();
+        let obj = Object { ctx: Some(()), destructor: destructor.clone() };
+        let destructor =
+            destructor.try_close().await.expect_err("Should have failed");
+        drop(obj);
+        assert!(destructor.try_close().await.is_ok());
+
+        // Dropping the detructor before the object is fine, if weird.
```
Comment (bnaecker) [inline illumos-utils/src/destructor.rs:376]: Spelling nit: "destructor"

**Review by bnaecker — APPROVED** [review]
Looks good, and great tests. I've a few small nits, but otherwise LGTM.

File: illumos-utils/src/link.rs (hunk 1 above)
Comment (smklein) [reply illumos-utils/src/link.rs]: Done

File: illumos-utils/src/link.rs (hunk 2 above)
Comment (smklein) [reply illumos-utils/src/link.rs]: Oooo, TIL - thanks! much better than manually making a string through `write`.

File: illumos-utils/src/destructor.rs (hunk 4 above)
Comment (smklein) [reply illumos-utils/src/destructor.rs]: Good call, updated

File: illumos-utils/src/destructor.rs (hunk 5 above)
Comment (smklein) [reply illumos-utils/src/destructor.rs]: Done

**Pull request merged**